src/Elastic.Apm/Metrics/Linux/GlobalMemoryStatus.cs

```diff
@@ -0,0 +1,186 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using Elastic.Apm.Logging;
+
+#if NET6_0_OR_GREATER
+using System.Buffers;
+using System.Buffers.Text;
+#endif
+
+namespace Elastic.Apm.Metrics.Linux
+{
+	internal static class GlobalMemoryStatus
+	{
+		public const string ProcMemInfo = "/proc/meminfo";
+
+#if NET6_0_OR_GREATER
+		private static readonly FileStreamOptions Options = new() { BufferSize = 0, Mode = FileMode.Open, Access = FileAccess.Read };
+		private static readonly byte Space = (byte)' ';
+
+		private static ReadOnlySpan<byte> MemAvailable => "MemAvailable:"u8;
+		private static ReadOnlySpan<byte> MemTotal => "MemTotal:"u8;
+		private static ReadOnlySpan<byte> KB => " kB"u8;
+#endif
+
+		public static (long totalMemory, long availableMemory) GetTotalAndAvailableSystemMemory(IApmLogger logger)
+			=> GetTotalAndAvailableSystemMemory(logger, null, false);
+
+		internal static (long totalMemory, long availableMemory) GetTotalAndAvailableSystemMemory(
+			IApmLogger logger, string pathPrefix, bool ignoreOs)
+		{
+			(long, long) failure = (-1, -1);
+			long totalMemory = -1, availableMemory = -1;
+
+			if (!RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && !ignoreOs)
+			{
+				logger.Trace()
+					?.Log("{ClassName} detected a non-Linux OS, therefore"
+						+ " proc/meminfo will not be reported", nameof(GlobalMemoryStatus));
+
+				return failure;
+			}
+
+			var memInfoPath = !string.IsNullOrEmpty(pathPrefix)
+				? Path.Combine(pathPrefix, ProcMemInfo.Substring(1))
+				: ProcMemInfo;
+
+			if (!File.Exists(memInfoPath))
+			{
+				logger.Error()?.Log("Unable to load memory information from {ProcMemInfo}", ProcMemInfo);
+				return failure;
+			}
+			try
+			{
+#if NET6_0_OR_GREATER
+				using var fs = new FileStream(memInfoPath, Options);
+				var buffer = ArrayPool<byte>.Shared.Rent(8192); // Should easily be large enough for max meminfo file.
+
+				try
+				{
+					var read = fs.Read(buffer);
```

Comment:
For future us can we add a few comments on what information from what format we are parsing from `/proc/meminfo`?

Reply:
Done